### PR TITLE
[Cost Insights] Add support for multiple types of entity cost breakdown

### LINKED ISSUE #3515: [RFC] Support for different breakdown dimensions in Cost Insights
**Status:** Closed for comments

## Need

Cost Insights supports a breakdown of entity costs by SKU as of #3268. We have a need internally at Spotify to show cost breakdowns by other dimensions - for example, by deployment (GCE vs GKE) or data pipeline trigger type (automatic, manual, backfill).

## Proposal

The modal dialog for SKU breakdowns should have tabs in cases where different dimensions are available. Alternatively, we can switch breakdowns to MUI's [Drawer](https://material-ui.com/components/drawers/) if this feels unappealing for a dialog.

In the current implementation, a singular breakdown is fetched as part of `CostInsightsApi.getProductInsights`, in `Entity.entities`. We could change the `entities` field to `Record<string, Entity[]>` to support multiple breakdowns by different types; this would require fetching all breakdowns for entities at load time. Alternatively, we can add an API method:

```typescript
type CostInsightsApi = {
  ...
  getEntityBreakdown(options: EntityInsightsOptions): Promise<Entity[]>
};

type EntityInsightsOptions = {
  entity: Entity;
  product: string;
  kind: string; // kind of breakdown, e.g. SKU, deployment
  duration: Duration;
  lastCompleteBillingDate: string;
};
```

In this case, the `Entity` type returned in `getProductInsights` should list the kinds of breakdowns supported by the entity.

A downside to this approach is that we won't know until fetching the breakdown whether the breakdown is meaningful. For example, do we want to show a deployment breakdown for things deployed in a single environment?

## Alternatives

One could argue that such breakdowns are another level of hierarchy in cost data, that an entity can have costs in different deployments which can _further_ be broken down by SKU. This would imply more of a drill-in hierarchical breakdown.

This could be useful in some examples and burdensome in others where switching between dimensions would be preferred. I'm not confident that we can (or want to) enforce that the relationship always be hierarchical. For our purposes internally, a drill-down approach seems undesirable since the dimensions offer a different insight than SKU.

Another alternative is to have a stacked bar chart, as shown in #3357. This seems undesirable in two ways:

- Breakdowns are available via two different mechanisms (hover + click), presenting a confusing UI.
- Limited number of breakdown entities supported

## Risks

Both API approaches detailed above are backwards-incompatible. We should be sure to note this with migration instructions in the changeset.

### PR DESCRIPTION
**Pull request by timbonicus**
Fixes #3515.

This updates the `Entity.entities` field from a single array of sub-entities to a keyed Record to allow multiple breakdown types (e.g. SKU versus deployment).

The root `Entity` returned by `CostInsightsApi.getProductInsights` is expected to only have a single breakdown (keyed by entity type, e.g. "service" or "pipeline"). These sub-entities can then have multiple breakdowns. These appear as tabs in the breakdown dialog:

entity-breakdown-dialog

I introduced the lightweight `pluralize` library, since we need rather broad pluralization capability for the breakdown keys.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/cost-insights/src/components/ProductInsightsCard/ProductEntityDialog.tsx [hunk 1]
```diff
@@ -14,179 +14,55 @@
...
+import React, { useState } from 'react';
+import { HeaderTabs } from '@backstage/core';
+import { Dialog, IconButton } from '@material-ui/core';
 import { default as CloseButton } from '@material-ui/icons/Close';
-import { CostGrowthIndicator } from '../CostGrowth';
-import { costFormatter, formatPercent } from '../../utils/formatters';
 import { useEntityDialogStyles as useStyles } from '../../utils/styles';
-import { BarChartOptions, Entity } from '../../types';
-
-function createRenderer(col: keyof RowData, classes: Record<string, string>) {
```
Comment (timbonicus) [inline plugins/cost-insights/src/components/ProductInsightsCard/ProductEntityDialog.tsx:27]: Moved the table to a separate `ProductEntityTable` component.

File: plugins/cost-insights/src/components/ProductInsightsCard/ProductEntityTable.tsx [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  options: ProductEntityTableOptions;
+};
+
+export const ProductEntityTable = ({
+  entityLabel,
+  entity,
+  options,
+}: ProductEntityTableProps) => {
+  const classes = useStyles();
+  const entities = entity.entities[entityLabel];
```
Comment (timbonicus) [inline plugins/cost-insights/src/components/ProductInsightsCard/ProductEntityTable.tsx:97]: Mostly just moved from `ProductEntityDialog`; minor functional change here.

File: plugins/cost-insights/src/utils/assert.ts [hunk 3]
```diff
@@ -50,3 +50,9 @@ export function findAlways<T>(
 ): T {
   return assertAlways(collection.find(callback));
 }
+
+export function findFirstKey<T>(
+  record: Record<string, T> | undefined,
+): string | undefined {
+  return Object.keys(record ?? {}).find(_ => true);
+}
```
Comment (ryanvazquez) [inline plugins/cost-insights/src/utils/assert.ts:58]: FYI, enumeration order of properties is a bit of a dicey subject and doesn't necessarily follow insertion order in some contexts.

https://stackoverflow.com/questions/5525795/does-javascript-guarantee-object-property-order/38218582#38218582

**Comment by ryanmvazquez** [comment]
Noticed in `ProductEntityDialog`, `M/M` - the last column header is fixed. We might want this header to change depending on period selected, esp since we're not supporting a MoM option now? Could even be something simpler like `+/-` ?

File: plugins/cost-insights/src/utils/grammar.ts [hunk 4]
```diff
@@ -22,20 +22,6 @@ const vowels = {
   u: 'U',
 };
 
-export const pluralOf = (
```
Comment (ryanmvazquez) [inline plugins/cost-insights/src/utils/grammar.ts:25]: 🔥 🔥 🔥

File: plugins/cost-insights/src/utils/assert.ts (hunk 3 above)
Comment (timbonicus) [reply plugins/cost-insights/src/utils/assert.ts:58]: The usage of this function expects to only find one key, so maybe `findAnyKey` would be a more appropriate name here. The point stands in the case of dialog tab order though, I use `Object.keys` similarly there. I'm not sure a good way around it though, the `entities` keys will likely be a cast object at runtime 🤔  It sounds at least like string keys may (coincidentally due to browser whims) preserve order.

Comment (ryanmvazquez) [reply plugins/cost-insights/src/utils/assert.ts:58]: Could enforce order by sorting lexicographically if we want to be super safe

Comment (timbonicus) [reply plugins/cost-insights/src/utils/assert.ts]: Eh, doesn't bother me 🤷‍♂️

**Review by ryanvazquez — APPROVED** [review]

File: plugins/cost-insights/src/types/Entity.ts [hunk 5]
```diff
@@ -31,8 +30,15 @@ export interface Entity {
   An entity could be atomic or composite. An atomic entity is indivisible
   and cannot be broken into sub-entities.
 
-  A composite entity can be broken down recursively into sub-entities
-  that generate cost **over the same time period**. All costs must sum to the root cost.
+  A composite entity is divided into sub-entities that account for portions
+  of the total cost **over the same time period**. The root entity is
+  expected to only have _one_ Record consisting of the sub-entities to display
+  in the product panel (keyed by the entity type, such as "service" for
```
Comment (brendasukh) [inline plugins/cost-insights/src/types/Entity.ts:36]: Missing closing parens

**Review by brendasukh — APPROVED** [review]

File: plugins/cost-insights/src/types/Entity.ts (hunk 5 above)
Comment (timbonicus) [reply plugins/cost-insights/src/types/Entity.ts:36]: I composed a commit-message haiku for this amazing spot ❤️

**Review by Rugvip — APPROVED** [review]

**Pull request merged**